Simplify Python code with some ruff rules SIM

cclauss opened the pull request:
https://docs.astral.sh/ruff/rules/#flake8-simplify-sim

Related to
* #13478

% `ruff check --select=SIM --ignore=SIM105,SIM108,SIM110,SIM112,SIM114,SIM910 --fix --unsafe-fixes`
```
Found 224 errors (139 fixed, 85 remaining).
```
% `ruff rule SIM118`
# in-dict-keys (SIM118)

Derived from the **flake8-simplify** linter.

Fix is always available.

## What it does
Checks for key-existence checks against `dict.keys()` calls.

## Why is this bad?
When checking for the existence of a key in a given dictionary, using
`key in dict` is more readable and efficient than `key in dict.keys()`,
while having the same semantics.

## Example
```python
key in foo.keys()
```

Use instead:
```python
key in foo
```

## Fix safety
Given `key in obj.keys()`, `obj` _could_ be a dictionary, or it could be
another type that defines a `.keys()` method. In the latter case, removing
the `.keys()` attribute could lead to a runtime error. The fix is marked
as safe when the type of `obj` is known to be a dictionary; otherwise, it
is marked as unsafe.

## References
- [Python documentation: Mapping Types](https://docs.python.org/3/library/stdtypes.html#mapping-types-dict)

[preview]: https://docs.astral.sh/ruff/preview/

eli-schwartz commented on unittests/linuxliketests.py:
```diff
@@ -1352,7 +1352,7 @@ def test_link_arg_fullname(self):
         #   see https://gitlab.freedesktop.org/pkg-config/pkg-config/-/blob/master/NEWS
         pkgconfigver = subprocess.check_output([PKG_CONFIG, '--version'])
-        if b'0.28' > pkgconfigver:
+        if pkgconfigver < b'0.28':
```
?

dcbaker replied:
I'm going to guess this is wanting to fix the [yoda condition](https://en.wikipedia.org/wiki/Yoda_conditions)

eli-schwartz reviewed (commented):
Most of these changes appear to be about inlining conditions onto a single line, which isn't immediately obvious is always superior. And it's not even mentioned in the PR description, let alone the commit message.

In general, the SIM rules are very opinionated stylistic things that have a good reason to not be enabled by default (or for flake8, even installed by default). I would rather have better justifications for using such rules...

dcbaker commented on mesonbuild/scripts/externalproject.py:
```diff
@@ -41,9 +41,7 @@ def write_stampfile(self) -> None:
-        if p.returncode == 0 and ('GNU Make' in o or 'waf' in o):
-            return True
-        return False
+        return bool(p.returncode == 0 and ('GNU Make' in o or 'waf' in o))
```
Do we really need `bool()` here? it looks like that would simplify to `bool and (bool or bool)`.

dcbaker commented on run_project_tests.py:
```diff
@@ -998,9 +993,7 @@ def have_working_compiler(lang: str, use_tmp: bool) -> bool:
-    if shutil.which('javac') and shutil.which('java'):
-        return True
-    return False
+    return bool(shutil.which('javac') and shutil.which('java'))
```
Do we actually need `bool()` here?

cclauss replied:
Yes, we do because the function type hint is a promise that we will return a `bool`.

Without `bool()` we would return a `str` on my Mac.

dcbaker replied:
Yeah, you're right.

eli-schwartz replied:
the `x and y` operator returns either x, or y, depending on whether x evaluates true. Same with `or`.

The `not` operator does convert to a guaranteed boolean.

dcbaker commented on run_project_tests.py:
```diff
@@ -1011,10 +1004,7 @@ def skip_dont_care(t: TestDef) -> bool:
-        return True
-
-    return False
+    return bool(mesonlib.is_osx() and '6 gettext' in str(t.path))
```
ditto

dcbaker commented on run_project_tests.py:
```diff
@@ -1060,17 +1050,12 @@ def should_skip_rust(backend: Backend) -> bool:
-        if has_broken_rustc():
-            return True
-    return False
+    return bool(mesonlib.is_windows() and has_broken_rustc())
```
ditto

cclauss replied:
This function is promised to return a `bool`.

Ruff lints superfast (1/3 of a secord for the entire CPython codebase) so it does not assume or validate that the retrun type of has_broken_rustc() `bool` or not.

dcbaker replied:
And both of the called functions promise to return a bool (and also do return a bool). I understand that ruff is not a static type checker, but if it's smart enough to notice that this function promises returns a bool, but not that the two called functions also promise to return a bool, that's surprising.

eli-schwartz replied:
I suspect ruff is actually a lot "dumber" (read: simple) and simply rewrites all code of the form:

```python
if cond():
    return BOOL
return OTHERBOOL
```
to:
```python
return BOOL if cond() else OTHERBOOL
```

And before that, it also rewrote:

```python
if cond1():
    if cond2():
        statement
```

to:
```python
if cond1() and cond2():
    statement
```

Ruff is doing code structure analysis, not type analysis, and not even return type analysis. It knows how to handle  builtin types, so that it can see the equivalence of `ret = bool(X)` and `if X: ret = True`.

dcbaker commented on run_project_tests.py:
```diff
@@ -1060,17 +1050,12 @@ def should_skip_rust(backend: Backend) -> bool:
-    if not shutil.which('wayland-scanner'):
-        return True
-    return False
+    return bool(not shutil.which('wayland-scanner'))
```
and here

dcbaker commented on unittests/machinefiletests.py:
```diff
@@ -94,10 +94,7 @@ def helper_create_binary_wrapper(self, binary, dir_=None, extra_args=None, **kwa
-            chbang = '#!/bin/env python3'
-        else:
-            chbang = '#!/usr/bin/env python3'
+        chbang = '#!/bin/env python3' if is_haiku() else '#!/usr/bin/env python3'
```
```suggestion
        chbang = '#!/usr/{} python3'.format('env' if is_haiku() else 'bin/env')
```

dcbaker commented:
I'm in favor of universally getting rid of bad uses of `.keys()` as well as replacing loops with `any`/`all`, but I agree wtih @eli-schwartz that not all of these changes (especially always using ternaries) is a blanket improvement. I'd also really like to see any of these that we're going to turn on have a lint that is permanently on in Meson, either through pylint or flake8

cclauss commented:
**Removed** `ruff rule SIM108`
# if-else-block-instead-of-if-exp (SIM108)

Derived from the **flake8-simplify** linter.

Fix is sometimes available.

## What it does
Check for `if`-`else`-blocks that can be replaced with a ternary operator.

## Why is this bad?
`if`-`else`-blocks that assign a value to a variable in both branches can
be expressed more concisely by using a ternary operator.

## Example
```python
if foo:
    bar = x
else:
    bar = y
```

Use instead:
```python
bar = x if foo else y
```

## References
- [Python documentation: Conditional expressions](https://docs.python.org/3/reference/expressions.html#conditional-expressions)

cclauss commented:
> I'd also really like to see any of these that we're going to turn on have a lint that is permanently on in Meson, either through pylint or flake8

The mods in this PR are code transformations that can happen once or they can be run on an ongoing basis.  The latter is more complex.  `flake8` plus https://pypi.org/project/flake8_simplify could be configured to lint the changes in this PR.

cclauss commented on mesonbuild/modules/java.py:
```diff
@@ -87,7 +87,7 @@ def __native_headers(self, state: ModuleState, args: T.Tuple[T.List[mesonlib.Fil
         ])
 
-        prefix = classes[0] if not package else package
+        prefix = package if package else classes[0]
```
```suggestion
        prefix = package or classes[0]
```

cclauss commented:
> I would rather have better justifications for using such rules...

Some of these rules like SIM118 (see commit message) are shorter and more efficient.  Others are just transforms that make the intent easier to understand with fewer lines of code (-77 lines in this PR).  The first line of the git commit contains a link to the list of all `SIM` rules and each of those is a link to a full page description of that rule.  The git commit also demonstrates `ruff rule SIM118` command which can provide useful on-the-fly docs.

As we saw with rule SIM108, some folks like certain transforms while others do not.  I have grown quite used to reading ternary-if and now just write them by habit.  Turning rules on or off is a choice for each project. 

https://docs.astral.sh/ruff
> Ruff can be used to replace [Flake8](https://pypi.org/project/flake8/) (plus dozens of plugins), [Black](https://github.com/psf/black), [isort](https://pypi.org/project/isort/), [pydocstyle](https://pypi.org/project/pydocstyle/), [pyupgrade](https://pypi.org/project/pyupgrade/), [autoflake](https://pypi.org/project/autoflake/), and more, all while executing tens or hundreds of times faster than any individual tool.

pylint lints with ruff and says in https://pylint.readthedocs.io/en/stable/#advised-linters-alongside-pylint
> ### Advised linters alongside pylint
> Projects that you might want to use alongside pylint include [ruff](https://github.com/charliermarsh/ruff) (really fast, with builtin auto-fix and a large number of checks taken from popular linters, but implemented in rust)

eli-schwartz commented:
Thanks for pointing all that out. All this information was also my starting premise.

My question was  not what does the PR do, but rather, whether each choice should be made... there's a reason we didn't simply install flake8-simplify.

P.S. I also know what ruff is! I package it for a linux distro, submit bug reports for it, and use it with meson for a couple things already. I don't need copious pages of copy-pasting snippets of ruff documentation to tell me what it does. And until ruff supports all the flake8 rules I'm not eager to switch. (Ruff initially didn't implement many rules because they suggested using black instead. Meson does not, and will not, use black because black produces terrible code. Even the rules that ruff has slowly added support for are mostly in preview.)

eli-schwartz commented on mesonbuild/compilers/mixins/emscripten.py:
```diff
@@ -78,9 +78,8 @@ def find_library(self, libname: str, env: 'Environment', extra_dirs: T.List[str]
-        if os.path.isabs(libname):
-            if os.path.exists(libname):
-                return [libname]
+        if os.path.isabs(libname) and os.path.exists(libname):
+            return [libname]
```
> make the intent easier to understand with fewer lines of code (-77 lines in this PR).

Simply reducing the number of lines is not the same as making the intent easier to understand. For an example, python supports rewriting this:

```suggestion
        if os.path.isabs(libname) and os.path.exists(libname): return [libname]
```

But most people wouldn't do that.

cclauss replied:
Let's not swim upstream against PEP8.

eli-schwartz commented on mesonbuild/modules/pkgconfig.py:
```diff
@@ -441,7 +441,7 @@ def _generate_pkgconfig_file(self, state: ModuleState, deps: DependenciesHelper,
                                  pkgroot: T.Optional[str] = None) -> None:
         coredata = state.environment.get_coredata()
         referenced_vars = set()
-        optnames = [x.name for x in BUILTIN_DIR_OPTIONS.keys()]
+        optnames = [x.name for x in BUILTIN_DIR_OPTIONS]
```
python also supports this:
```suggestion
        referenced_vars = set(); optnames = [x.name for x in BUILTIN_DIR_OPTIONS]
```

It, too, is fewer lines. It, too, is widely considered less readable.

cclauss replied:
PEP8

eli-schwartz commented:
I'm trying to point out a general principle here, that quoting dogma is not the same as stating a reason. But I suspect the disconnect here may be part and parcel of the principle that:
> because an automated linter includes the selectable option to say you CAN do it, therefore you SHOULD do it

eli-schwartz commented on mesonbuild/utils/universal.py:
```diff
@@ -1876,10 +1873,7 @@ def detect_subprojects(spdir_name: str, current_dir: str = '',
-        if substr in s:
-            return True
-    return False
+    return any(substr in s for s in strlist)
```
IIRC these any() approaches are much slower. It's an interesting quirk of a low-level function whose body is basically entirely defined by either a short-circuiting for loop or a short-circuiting `any()`.

cclauss replied:
The ruff `Why is this bad?` section says it is faster but perhaps you have a benchmark that proves it wrong.

% [`ruff rule SIM110`](https://docs.astral.sh/ruff/rules/reimplemented-builtin/)
# reimplemented-builtin (SIM110)

## Why is this bad?
Using a builtin function is more concise and readable. Builtins are also
more efficient than `for` loops.

---

I am OK to turn off `ruff rule SIM110` if you think that is best.  I find that when I read the code out loud, the use of `any()` makes the code read the way I think.  Others might disagree.

dcbaker replied:
It's really unfortunate because I think that `any()`/`all()` is **_MUCH_** more readable than the for loop, and the vertical space saving is very nice. However, any seems to be about 1/3 as fast (Python 3.11.9, NixOS):

```python
loop = 'abcdef'.split()
found = 'f'
nfound = 'g'


def test1():
    for x in loop:
        if x == found:
            return True
    return False


def test2():
    return any(x == found for x in loop)


def test3():
    for x in loop:
        if x == nfound:
            return True
    return False


def test4():
    return any(x == nfound for x in loop)

if __name__ == "__main__":
    import timeit

    print('for loop (found)    :', timeit.timeit(test1))
    print('for loop (not found):', timeit.timeit(test3))
    print('any() (found)       :', timeit.timeit(test2))
    print('any() (not found)   :', timeit.timeit(test4))
```

```
for loop (found)    : 0.051076093994197436
for loop (not found): 0.04388196699437685
any() (found)       : 0.15422860698890872
any() (not found)   : 0.15568504799739458
```

cclauss replied:
On Python 3.13 release candidate 1 on M1 Mac:
```
for loop (found)    : 0.09455837499990594
for loop (not found): 0.09435533400028362
any() (found)       : 0.27777291600068565
any() (not found)   : 0.27852720800001407
```

Should we drop SIM110 ?

cclauss commented:
> until ruff supports all the flake8 rules I'm not eager to switch.

I am not advocating that you switch.  Ruff is not yet at v1.  flake8 and pylint continue to be adequate.  This PR is about looking at a few simplifying code transformations.  If a rule does not improve the code then we drop it.  If none of the rules improve the code then we close the PR.

> (Ruff initially didn't implement many rules because they suggested using black instead. Meson does not, and will not, use black because black produces terrible code. Even the rules that ruff has slowly added support for are mostly in preview.)

I recognize that you wish to format code by hand so I am not suggesting that you drop flake8 or pylint.  I am not aware of any flake8 rules that ruff does not support except the [Exxx rules](https://docs.astral.sh/ruff/rules/#error-e) in preview that are better handled by a code formatter.  I don't think that the Ruff team spends time convincing people to use psf/black anymore given that they now have `ruff format`.

> quoting dogma is not the same as stating a reason.

This is computer science, not religion or politics.  The `Why is this bad?` section for each rule is a curated explanation.  If the explanation does not justify the change then we should drop that rule.

> because an automated linter includes the selectable option to say you CAN do it, therefore you SHOULD do it

I agree 100%.  The initial commit `--ignored` five SIM rules and through our discussions, we have added a sixth.

cclauss commented:
Removed all instances of SIM110 because it is less efficient than the code it releases.

Added a pull request to amend the Ruff docs at https://github.com/astral-sh/docs/pulls
